ext/src/js/modules/document.js

```diff
@@ -0,0 +1,173 @@
+import { downloadFile } from './file-download';
+import { sanitizeCSS } from './sanitize';
+
+class Document {
+    constructor(html, url, doctype, htmlAttributes) {
+        this.url = url;
+        this.doctype = doctype;
+        this.dom = document.createElement('html');
+        this.iframes = [];
+        this.styleIndex = 0;
+        this.styleNodes = new Map();
+        this.favicon = null;
+        this.dom.innerHTML = html;
+        for (const k in htmlAttributes) {
+            this.dom.setAttribute(k, htmlAttributes[k]);
+        }
+        this.nodeTransformFunctons = new Map([
+            ['SCRIPT', (node) => node.remove()],
+            ['LINK', this.transformLink],
+            ['STYLE', this.transformStyle],
+            ['IMG', this.transfromImg],
+            ['IFRAME', this.transfromIframe],
+            ['BASE', this.setUrl]
+        ]);
+    }
+
+    absoluteUrl(url) {
+        if (!url) {
+            return this.url;
+        }
+        return new URL(url, this.url).href;
+    }
+
+    getDomAsText() {
+        return `${this.doctype}${this.dom.outerHTML}`;
+    }
+
+    async transformDom() {
+        await this.walkDOM(this.dom);
+        await this.setStyleNodes();
+        if (!this.favicon) {
+            this.favicon = await downloadFile(this.absoluteUrl('/favicon.ico'));
+            if (this.favicon) {
+                const faviconElement = document.createElement('link');
+                faviconElement.setAttribute('rel', 'icon');
+                faviconElement.setAttribute('href', this.favicon);
+                this.dom.getElementsByTagName('head')[0].appendChild(faviconElement);
+            }
+        }
+    }
+
+    async walkDOM(node) {
+        await this.transformNode(node);
+        const children = [...node.childNodes];
+        return Promise.allSettled(children.map(async (node) => {
+            await this.walkDOM(node).catch(e => console.log("Error while transforming DOM: ", e));
+        }));
+    }
+
+    async setStyleNodes() {
+        const sortedStyles = new Map([...this.styleNodes.entries()].sort((e1, e2) => e1[0] - e2[0]));
+        let parent;
+        if (this.dom.getElementsByTagName("head")) {
+            parent = this.dom.getElementsByTagName("head")[0];
+        } else {
+            parent = this.dom.documentElement;
+        }
+        sortedStyles.forEach(style => {
+            parent.appendChild(style);
+        });
+    }
+
+    async transformNode(node) {
+        if (node.nodeType !== Node.ELEMENT_NODE) {
+            return;
+        }
+        const transformFunction = this.nodeTransformFunctons.get(node.nodeName);
+        await this.rewriteAttributes(node);
+        if (transformFunction) {
+            // todo bind vs apply
```

Comment:
Can you change this to transformFunction.call(this, node)
(and remove all the doc input params from transform functions)